src/lib/extension-load-queue.ts

```diff
@@ -0,0 +1,118 @@
+import {useEffect, useState} from 'react';
+
+declare global {
+    interface Window {
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        [key: symbol]: any;
+        QUEUE_SYMBOL: boolean;
+    }
+}
+
+export type ControllerLoadedCallback<T> = (controller: T) => void;
+
+export const QUEUE_SYMBOL = Symbol.for('queue');
+
+export type ScriptStore<T> = ControllerLoadedCallback<T>[];
+
+export interface CreateLoadQueueArgs<T> {
+    store: ScriptStore<T>;
+    createController: () => T;
+    isQueueCreated?: boolean;
+    onQueueCreated?: (created: boolean) => void;
+}
+
+// TODO: this function is a very frequently used utility,
+//  so we should place it in a separate top-level file and document it.
+export const isBrowser = () => typeof window !== 'undefined' && typeof document !== 'undefined';
+
+export const getScriptStore = <T>(key: symbol): ScriptStore<T> => {
+    if (isBrowser()) {
+        (window as Window)[key] = (window as Window)[key] || [];
+        return (window as Window)[key];
+    } else {
+        throw new Error('This functionality should be employed on the client-side.');
+    }
+};
+
+export const getQueueStore = () => {
+    if (isBrowser()) {
+        (window as Window)[QUEUE_SYMBOL] = (window as Window)[QUEUE_SYMBOL] || false;
```

Comment:
`as Window` casts are weird — did you encounter any issues with declaration merging?